Source/Core/Core/DSP/Jit/x64/DSPJitTables.cpp

```diff
@@ -0,0 +1,386 @@
+  {0xa400, 0xe600, &DSPEmitter::mulxac},
+  {0xa600, 0xe600, &DSPEmitter::mulxmv},
+  {0xb100, 0xf700, &DSPEmitter::tst},
+
+  // C-D
+  {0xc000, 0xe700, &DSPEmitter::mulc},
+  {0xc100, 0xe700, &DSPEmitter::cmpar},
+  {0xc200, 0xe600, &DSPEmitter::mulcmvz},
+  {0xc400, 0xe600, &DSPEmitter::mulcac},
+  {0xc600, 0xe600, &DSPEmitter::mulcmv},
+
+  // E
+  {0xe000, 0xfc00, &DSPEmitter::maddx},
+  {0xe400, 0xfc00, &DSPEmitter::msubx},
+  {0xe800, 0xfc00, &DSPEmitter::maddc},
+  {0xec00, 0xfc00, &DSPEmitter::msubc},
+
+  // F
+  {0xf000, 0xfe00, &DSPEmitter::lsl16},
+  {0xf200, 0xfe00, &DSPEmitter::madd},
+  {0xf400, 0xfe00, &DSPEmitter::lsr16},
+  {0xf600, 0xfe00, &DSPEmitter::msub},
+  {0xf800, 0xfc00, &DSPEmitter::addpaxz},
+  {0xfc00, 0xfe00, &DSPEmitter::clrl},
+  {0xfe00, 0xfe00, &DSPEmitter::movpz},
+}};
+
+constexpr std::array<JITOpInfo, 25> s_opcodes_ext
+{{
+  {0x0000, 0x00fc, &DSPEmitter::nop},
+
+  {0x0004, 0x00fc, &DSPEmitter::dr},
+  {0x0008, 0x00fc, &DSPEmitter::ir},
+  {0x000c, 0x00fc, &DSPEmitter::nr},
+  {0x0010, 0x00f0, &DSPEmitter::mv},
+
+  {0x0020, 0x00e4, &DSPEmitter::s},
+  {0x0024, 0x00e4, &DSPEmitter::sn},
+
+  {0x0040, 0x00c4, &DSPEmitter::l},
+  {0x0044, 0x00c4, &DSPEmitter::ln},
+
+  {0x0080, 0x00ce, &DSPEmitter::ls},
+  {0x0082, 0x00ce, &DSPEmitter::sl},
+  {0x0084, 0x00ce, &DSPEmitter::lsn},
+  {0x0086, 0x00ce, &DSPEmitter::sln},
+  {0x0088, 0x00ce, &DSPEmitter::lsm},
+  {0x008a, 0x00ce, &DSPEmitter::slm},
+  {0x008c, 0x00ce, &DSPEmitter::lsnm},
+  {0x008e, 0x00ce, &DSPEmitter::slnm},
+
+  {0x00c3, 0x00cf, &DSPEmitter::ldax},
+  {0x00c7, 0x00cf, &DSPEmitter::ldaxn},
+  {0x00cb, 0x00cf, &DSPEmitter::ldaxm},
+  {0x00cf, 0x00cf, &DSPEmitter::ldaxnm},
+
+  {0x00c0, 0x00cc, &DSPEmitter::ld},
+  {0x00c4, 0x00cc, &DSPEmitter::ldn},
+  {0x00c8, 0x00cc, &DSPEmitter::ldm},
+  {0x00cc, 0x00cc, &DSPEmitter::ldnm},
+}};
+// clang-format on
+
+namespace
+{
+std::array<JITFunction, 65536> s_op_table;
+std::array<JITFunction, 256> s_ext_op_table;
+bool s_tables_initialized = false;
+
+template <size_t N>
+auto FindByOpcode(UDSPInstruction opcode, const std::array<JITOpInfo, N>& data)
+{
+  return std::find_if(data.cbegin(), data.cend(),
+                      [opcode](const auto& info) { return (opcode & info.mask) == info.opcode; });
+}
+}  // Anonymous namespace
+
+JITFunction GetOp(UDSPInstruction inst)
+{
+  return s_op_table[inst];
+}
+
+JITFunction GetExtOp(UDSPInstruction inst)
+{
+  const bool has_seven_bit_extension = (inst >> 12) == 0x3;
+
+  if (has_seven_bit_extension)
+    return s_ext_op_table[inst & 0x7F];
+
+  return s_ext_op_table[inst & 0xFF];
+}
+
+void InitInstructionTables()
+{
+  if (s_tables_initialized)
+    return;
+
+  // ext op table
+  for (size_t i = 0; i < s_ext_op_table.size(); i++)
+  {
+    const auto iter = FindByOpcode(static_cast<UDSPInstruction>(i), s_opcodes_ext);
+    if (iter == s_opcodes_ext.cend())
+      continue;
+
+    if (s_ext_op_table[i] == nullptr)
```

Comment:
Are you relying here that the static array is initialized with zero? This is valid, but IMO error prone if someone wants to move it to the heap. May you set s_ext_op_table[i] = nullptr; at the top of this loop and drop this branch here?

Reply:
Why would heap allocating ever be a good idea in this scenario?